File: Programmers/hyeonyoung/다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights)
{
    int now = 0, weight_sum = 0;
    queue<pair<int, int>> Q; // weight, time

    for (int x : truck_weights)
    {
        now++;
        if (now == Q.front().second)
        {
            weight_sum -= Q.front().first;
            Q.pop();
        }
        while (weight_sum + x > weight)
        {
            weight_sum -= Q.front().first;
            now = Q.front().second;
            Q.pop();
        }
        weight_sum += x;
        Q.push({x, now + bridge_length});
    }
    while (!Q.empty())
    {
        weight_sum -= Q.front().first;
        now = Q.front().second;
        Q.pop();
    }

    return now;
}

File: Programmers/sohui/다리를 지나는 트럭.py
from collections import deque


def solution(bridge_length, weight, truck_weights):
    truck_weights = deque(truck_weights)
    bridge = deque([0 for i in range(bridge_length)])
    time = 0
    bridge_weight = 0
    while len(bridge) != 0:
        out = bridge.popleft()
        bridge_weight -= out
        time += 1

        if truck_weights:
            if bridge_weight + truck_weights[0] <= weight:
                left = truck_weights.popleft()
                bridge_weight += left
                bridge.append(left)
            else:
                bridge.append(0)
    return time